### ISSUE TITLE
Generic OAuth via OIDC: Clarification/Issue with fix for sub support

### ISSUE THREAD

**Issue by aval13** [grafana/grafana#79328]
### What happened?

Hello,
There was an issue opened some time ago about being able to use the sub field from OIDC as a valid identifier for login using Generic Auth after a security fix which broke this functionality.
Issue: https://github.com/grafana/grafana/issues/64507
The fix was to consider the sub field as authid and based on the debug logs it seems to work as expected, but we have encountered a very strange behaviour:
- with allow_sign_up: false the authentication fails with message
"Not allowing login, user not found in internal user database and allow signup = false"
- with allow_sign_up: true the authentication fails with message
"user already exists"

Please see below for more information

### What did you expect to happen?

Since all the fields are properly parsed and extracted from the OAuth data and everything looks exactly like we expect it to look, I would expect the login process to identify the user and allow the login.

### Did this work before?

We did not have time to test the fix once it was released, so I do not know.
We tested with 9.5.8 (the first 9.5.x to receive the fix), 9.5.14 and 10.2.2, same behaviour.

### How do we reproduce it?

1. Deploy a Grafana 9.5.14 (latest 9.5.x) and configure generic oauth and allow_sign_up on false (see below)
2. Provision a user using Grafana API (see below)
3. Try to login

Grafana configuration, with allow sign up on false, switch it to true for the other case:
```
    auth:
      oauth_skip_org_role_update_sync: true
    auth.generic_oauth:
      enabled: true
      client_id: <OIDC_CLIENT_ID>
      client_secret: <OIDC_CLIENT_SECRET>
      scopes: "openid email profile"
      auth_url: <AUTH_URL>
      token_url: <TOKEN_URL>
      api_url: <API_URL>
      allow_sign_up: false
      name: "OIDC Login"
      login_attribute_path: sub
      name_attribute_path: name
      email_attribute_name: email
      skip_org_role_sync: true
```

Using a sub of the form "abcd1234", we provision an user with "dummy placeholders".
After provision, curl on /api/users/lookup?loginOrEmail=abcd1234' | jq .
```
{
  "id": 3,
  "email": "abcd1234@localhost",
  "name": "abcd1234",
  "login": "abcd1234",
  "theme": "",
  "orgId": 1,
  "isGrafanaAdmin": false,
  "isDisabled": false,
  "isExternal": false,
  "isExternallySynced": false,
  "authLabels": null,
  "updatedAt": "2023-12-11T14:23:14Z",
  "createdAt": "2023-12-11T14:23:14Z",
  "avatarUrl": ""
}
```

Logs in debug with allow_sign_up on false:
```
logger=oauth t=2023-12-11T14:30:39.763914836Z level=debug msg="OAuthLogin: got token" expiry="2023-12-11 15:30:39.763880604 +0000 UTC m=+4243.500602242" type=Bearer has_refresh_token=false
logger=oauth.generic_oauth t=2023-12-11T14:30:39.763944538Z level=debug msg="Getting user info"
logger=oauth.generic_oauth t=2023-12-11T14:30:39.763951493Z level=debug msg="Extracting user info from OAuth token"
logger=oauth.generic_oauth t=2023-12-11T14:30:39.764091214Z level=debug msg="Received id_token" raw_json="{\"exp\":1702308639,\"at_hash\":\"REDACTED\",\"sub\":\"abcd1234\",\"iat\":1702305039,\"iss\":\"https://REDACTED\",\"acr\":\"loa-4\",\"aud\":[\"REDACTED\"],\"azp\":\"REDACTED\",\"auth_time\":1702296188}" data="Name: , Displayname: , Login: , Username: , Email: , Upn: , Attributes: map[]"
logger=oauth.generic_oauth t=2023-12-11T14:30:39.764110472Z level=debug msg="Getting user info from API"
logger=oauth.generic_oauth t=2023-12-11T14:30:39.940059733Z level=debug msg="HTTP GET" url=https://REDACTED/userinfo status="200 OK" response_body="{\"family_name\":\"FAMILY\",\"email\":\"correct.email@domain.com\",\"name\":\"NAME FAMILY\",\"given_name\":\"NAME\",\"sub\":\"abcd1234\"}"
logger=oauth.generic_oauth t=2023-12-11T14:30:39.94011461Z level=debug msg="Received user info response from API" raw_json="{\"family_name\":\"FAMILY\",\"email\":\"correct.email@domain.com\",\"name\":\"NAME FAMILY\",\"given_name\":\"NAME\",\"sub\":\"abcd1234\"}" data="Name: NAME FAMILY, Displayname: , Login: , Username: , Email: correct.email@domain.com, Upn: , Attributes: map[]"
logger=oauth.generic_oauth t=2023-12-11T14:30:39.940124734Z level=debug msg="Processing external user info" source=token data="Name: , Displayname: , Login: , Username: , Email: , Upn: , Attributes: map[]"
logger=oauth.generic_oauth t=2023-12-11T14:30:39.940186068Z level=debug msg="Unable to find user info name"
logger=oauth.generic_oauth t=2023-12-11T14:30:39.940198178Z level=debug msg="Searching for login among JSON" loginAttributePath=sub
logger=oauth.generic_oauth t=2023-12-11T14:30:39.940223831Z level=debug msg="Processing external user info" source=API data="Name: NAME FAMILY, Displayname: , Login: , Username: , Email: correct.email@domain.com, Upn: , Attributes: map[]"
logger=oauth.generic_oauth t=2023-12-11T14:30:39.940248176Z level=debug msg="Setting user info name from nameAttributePath" nameAttributePath=name
logger=oauth.generic_oauth t=2023-12-11T14:30:39.940253743Z level=debug msg="Set user info email from extracted email" email=correct.email@domain.com
logger=oauth.generic_oauth t=2023-12-11T14:30:39.940258964Z level=debug msg="User info result" result="Id: abcd1234, Name: NAME FAMILY, Email: correct.email@domain.com, Login: abcd1234, Role: , Groups: []"
logger=oauth t=2023-12-11T14:30:39.940278454Z level=debug msg="OAuthLogin got user info" userInfo="Id: abcd1234, Name: NAME FAMILY, Email: correct.email@domain.com, Login: abcd1234, Role: , Groups: []"
logger=oauth t=2023-12-11T14:30:39.940285044Z level=debug msg="Building external user info from OAuth user info"
logger=oauth t=2023-12-11T14:30:39.940290723Z level=debug msg="Syncing Grafana user with corresponding OAuth profile"
logger=context userId=0 orgId=0 uname= t=2023-12-11T14:30:39.940781272Z level=warn msg="Not allowing login, user not found in internal user database and allow signup = false" authmode=oauth_generic_oauth
logger=context userId=0 orgId=0 uname= t=2023-12-11T14:30:39.94080539Z level=warn msg="system administrator has disabled signup"
```

Logs in debug with allow_sign_up on true:
```
logger=oauth t=2023-12-11T14:30:57.519236918Z level=debug msg="OAuthLogin: got token" expiry="2023-12-11 15:30:57.519205729 +0000 UTC m=+3905.680596644" type=Bearer has_refresh_token=false
logger=oauth.generic_oauth t=2023-12-11T14:30:57.51926787Z level=debug msg="Getting user info"
logger=oauth.generic_oauth t=2023-12-11T14:30:57.519274243Z level=debug msg="Extracting user info from OAuth token"
logger=oauth.generic_oauth t=2023-12-11T14:30:57.519420318Z level=debug msg="Received id_token" raw_json="{\"azp\":\"REDACTED\",\"auth_time\":1702296188,\"aud\":[\"REDACTED\"],\"acr\":\"loa-4\",\"sub\":\"abcd1234\",\"iss\":\"https://auth.REDACTED\",\"iat\":1702305057,\"exp\":1702308657,\"at_hash\":\"REDACTED\"}" data="Name: , Displayname: , Login: , Username: , Email: , Upn: , Attributes: map[]"
logger=oauth.generic_oauth t=2023-12-11T14:30:57.519436316Z level=debug msg="Getting user info from API"
logger=oauth.generic_oauth t=2023-12-11T14:30:57.689502168Z level=debug msg="HTTP GET" url=https://REDACTED/userinfo status="200 OK" response_body="{\"email\":\"correct.email@domain.com\",\"given_name\":\"NAME\",\"family_name\":\"FAMILY\",\"sub\":\"abcd1234\",\"name\":\"NAME FAMILY\"}"
logger=oauth.generic_oauth t=2023-12-11T14:30:57.689555472Z level=debug msg="Received user info response from API" raw_json="{\"email\":\"correct.email@domain.com\",\"given_name\":\"NAME\",\"family_name\":\"FAMILY\",\"sub\":\"abcd1234\",\"name\":\"NAME FAMILY\"}" data="Name: NAME FAMILY, Displayname: , Login: , Username: , Email: correct.email@domain.com, Upn: , Attributes: map[]"
logger=oauth.generic_oauth t=2023-12-11T14:30:57.689565081Z level=debug msg="Processing external user info" source=token data="Name: , Displayname: , Login: , Username: , Email: , Upn: , Attributes: map[]"
logger=oauth.generic_oauth t=2023-12-11T14:30:57.689607924Z level=debug msg="Unable to find user info name"
logger=oauth.generic_oauth t=2023-12-11T14:30:57.689615481Z level=debug msg="Searching for login among JSON" loginAttributePath=sub
logger=oauth.generic_oauth t=2023-12-11T14:30:57.689641226Z level=debug msg="Processing external user info" source=API data="Name: NAME FAMILY, Displayname: , Login: , Username: , Email: correct.email@domain.com, Upn: , Attributes: map[]"
logger=oauth.generic_oauth t=2023-12-11T14:30:57.689662051Z level=debug msg="Setting user info name from nameAttributePath" nameAttributePath=name
logger=oauth.generic_oauth t=2023-12-11T14:30:57.689672312Z level=debug msg="Set user info email from extracted email" email=correct.email@domain.com
logger=oauth.generic_oauth t=2023-12-11T14:30:57.689678403Z level=debug msg="User info result" result="Id: abcd1234, Name: NAME FAMILY, Email: correct.email@domain.com, Login: abcd1234, Role: , Groups: []"
logger=oauth t=2023-12-11T14:30:57.689692894Z level=debug msg="OAuthLogin got user info" userInfo="Id: abcd1234, Name: NAME FAMILY, Email: correct.email@domain.com, Login: abcd1234, Role: , Groups: []"
logger=oauth t=2023-12-11T14:30:57.689699647Z level=debug msg="Building external user info from OAuth user info"
logger=oauth t=2023-12-11T14:30:57.689705229Z level=debug msg="Syncing Grafana user with corresponding OAuth profile"
logger=context userId=0 orgId=0 uname= t=2023-12-11T14:30:57.691335141Z level=warn msg="user already exists"
```

Basically the logs are identical, except once it complains it cannot find the user, the other time it complains the user already exists.

Please let us know if we are doing something wrong.

Thank you.

### Environment (with versions)?

Grafana: 9.5.14 / 10.2.2
OS:
Browser:

### Grafana platform?

Kubernetes

**Comment by aval13** [comment]
Hello?

Can anybody shed any light on this?

Thank you

**Comment by aval13** [comment]
Hello again,

@Jguer , @tonypowa , @kalleep , could you please help with any insights on this?

Thank you

**Comment by Jguer** [comment]
Hi @aval13 , if the user has been created with the Grafana API and allow signup and insecure_email_lookup and disabled then the user will be indeed unable to login.

**Comment by Jguer** [comment]
I do see how `allow_sign_up` doesn't behave how you'd expect it.

It does not allow for users to be added in other ways and then authed with oauth but instead stops any new user from oauth from registering

**Comment by aval13** [comment]
> Hi @aval13 , if the user has been created with the Grafana API and allow signup and insecure_email_lookup and disabled then the user will be indeed unable to login.

What I want to use and tested:
auth . generic_oauth . allow_sign_up is **false**
auth . oauth_allow_insecure_email_lookup is **true**

> I do see how allow_sign_up doesn't behave how you'd expect it.
> It does not allow for users to be added in other ways and then authed with oauth but instead stops any new user from oauth from logging in

Yes and no.

If I provision the user via Grafana API using their email as value for login, email and name, the login is successful.
If I provision the same way but using their sub (ok, the email is sub@localhost), I get the behaviour above.

So using the email match "route", all ok.
Using the sub, at the step of
"Syncing Grafana user with corresponding OAuth profile"
I get the contradiction between

allow_sign_up **false** - Grafana log: "Not allowing login, user not found in internal user database and allow signup = false"
User receives message in the login page:
Login Failed
system administrator has disabled signup

allow_sign_up **true** - Grafana log: "user already exists"
User receives message in the login page:
Login Failed
user already exists

It looks to me like, when trying to identify the local Grafana user using the data from oauth, based on the allow sign up being true or false, for some reason there are different checks to verify if the user exists or not.
How can it be possible that with the same oauth data and the same local user provisioned in the same way, once it says "hey, there is no local user", the other time says "hey, the user already exists"

I did another test:
I put allow_sign_up on true, did not provision the user, the login worked normally.
Comparing the user I provisioned with the user created by Grafana based on the signup and the oauth data, the jsons returned by the Grafana API are identical.

And one more test:
the same as above, but this time with oauth_allow_insecure_email_lookup on false, still works.

To sum it up what I tested so far:

auth . oauth_allow_insecure_email_lookup = true
auth . generic_oauth . allow_sign_up = false
user provisioned with email -> works (OK)
user provisioned with sub -> failed: user not found in internal user database (ISSUE)

auth . oauth_allow_insecure_email_lookup = true
auth . generic_oauth . allow_sign_up = true
user provisioned with email -> works (OK?)
user provisioned with sub -> failed: user already exists (ISSUE)

auth . oauth_allow_insecure_email_lookup = false
auth . generic_oauth . allow_sign_up = true
user provisioned with email -> failed: user already exists (?)
user provisioned with sub -> failed: user already exists (?)

auth . oauth_allow_insecure_email_lookup = false
auth . generic_oauth . allow_sign_up = false
user provisioned with email -> failed: user not found in internal user database (OK)
user provisioned with sub -> failed: user not found in internal user database (ISSUE?)

**Comment by Jguer** [comment]
Indeed provisioning a user with `sub` is not possible as the sub is not stored in the database's user table but in a separate table that has no API support.

from the scenarios described I don't see any issue

**Comment by aval13** [comment]
Hello again

> Indeed provisioning a user with `sub` is not possible as the sub is not stored in the database's user table but in a separate table that has no API support.
> from the scenarios described I don't see any issue

Ok, so this means I must have missread/missunderstood your comment here
https://github.com/grafana/grafana/issues/64507#issuecomment-1466148736
the part with
"we might have something better. We do recognize the sub claim as being unique not only in your system but also in the OIDC standard [...] and as such we could use it for **matching directly** (without bringing username into the question)"

As a side note for anybody that might run into the same issue as me, the "workaround" would be that besides the provisioning of the user via Grafana API, a query must also be run in order to add an entry in the user_auth table to make things work, something like:
INSERT INTO user_auth VALUES (NULL, <REPLACE_ID_OF_THE_PROVISIONED_USER>, 'oauth_generic_oauth', '<REPLACE_SUB>', DATETIME(), NULL, NULL, NULL, NULL, NULL);

I tested the above workaround, now I can properly login using only the sub.

To finalize this issue, to sum it up:
If I want to use generic_oauth with OIDC **AND** use allow_sign_up on **FALSE** to limit access to explictly provided users:
- if I provision an user using the email as login, email and name, and set auth . oauth_allow_insecure_email_lookup on true, it works
- if I **NEED** to provision using the sub, there is no supported way by Grafana to do this, except for the workaround above.

@Jguer Is this conclusion correct?

Thank you

**Comment by Jguer** [comment]
:+1: @aval13 , for now that would work if you want pre-provisioning of oauth users instead of JIT creation. We're evaluating how we can improve this
